nova/modules/core/libraries/Nova_event.php

```diff
@@ -0,0 +1,133 @@
+    $ref =& $this->callbacks;
+    
+    while(true)
+    {
+      // Add each callback in reverse order, so when reversed again given that 
+      // $callbackStack is LIFO, the first one in this array (so the first one 
+      // added) will be executed first.
+      if(!empty($ref['_callbacks']))
+      {
+        foreach(array_reverse($ref['_callbacks']) as $callback)
+        {
+          $callbackStack[] = $callback;
+        }
+      }
+      
+      $seg = array_shift($event);
+      
+      // Terminate the loop if there's no events specified under the next $seg
+      // key or if there are no more segments left to traverse.
+      if(!$seg || !isset($ref['_children'][$seg]))
+      {
+        break;
+      }
+      
+      // Change the pointer to the next segment.
+      $ref =& $ref['_children'][$seg];
+    };
+    
+    // Send the name of the event with the special key _ci_event in case a 
+    // listener is defined across a namespace (as opposed to just a terminating 
+    // event) yet still wants to isolate is execution to only some subset of
+    // events.
+    $eventData['_ci_event'] = $event;
+    
+    // $callbackStack is arranged in LIFO, so array_reverse for iteration
+    foreach(array_reverse($callbackStack) as $callback)
+    {
+      $callback($eventData);
+    }
+  }
+  
+  /**
+   * Register a callback for the event segments specified as $event. This 
+   * callback will be executed whenever {@see trigger()} is called for an
+   * event with all the same segments.
+   *
+   * A listener may be defined for an $event that is only some of the segments
+   * specified in $event from trigger(). If multiple listeners have been 
+   * attached, they will execute in order of most-specific to least-specific.
+   *
+   * Given these callbacks:
+   *
+   *      $this->event->listen(['location', 'view', 'output'], function($event){
+   *        $event['output'] .= 'c';
+   *      });
+   *      $this->event->listen(['location', 'view', 'output'], function($event){
+   *        $event['output'] .= 'd';
+   *      });
+   *  
+   *      $this->event->listen(['location', 'view', 'output', 'main', 'index'], 
+   *          function($event){
+   *        $event['output'] .= 'a';
+   *      });
+   *  
+   *      $this->event->listen(['location', 'view', 'output', 'main', 'index'], 
+   *          function($event){
+   *        $event['output'] .= 'b';
+   *      });
+   * 
+   * The following string will be added to the output:
+   *
+   *      abcd
+   *
+   * @param string[] $event name segmemts (eg: ['location','view','data'])
+   * @param callback $callback to be invoked when matching event is triggered
+   */
+  public function listen($event, $callback)
+  {  
+    $ref =& $this->callbacks;
+    
+    while($seg = array_shift($event))
+    {
+      if(!isset($ref['_children'][$seg]))
+      {
+        $ref['_children'][$seg] = [
+          '_children' => [],
+          '_callbacks' => []
```

Comment:
What are these `callbacks`? Aren't we added a callback? Is this the ability to continually nest callbacks?

Reply:
These are the callbacks at this event depth. My comment for `_children` should explain this, but I will also add a comment for this to explain it.